csharp/src/Microsoft.ML.OnnxRuntime/OnnxRuntime.shared.cs

```diff
@@ -52,50 +64,77 @@ public enum OrtLanguageProjection
     /// </summary>
     public sealed class OrtEnv : SafeHandle
     {
-        private static readonly Lazy<OrtEnv> _instance = new Lazy<OrtEnv>(() => new OrtEnv());
+        private static readonly Lazy<OrtEnv> _instance = new Lazy<OrtEnv>(CreateDefaultOrtEnv);
         private static LogLevel envLogLevel = LogLevel.Warning;
+        private readonly bool _owned;
```

Comment:
>donly bool _owned; [](http://example.com/codeflow?start=19&length=18)

I will come back to this PR, but it is important to understand, that on the library level there is only one refcounted instance of the environment. It is not possible to create it more than once. This has the following implications:
1) The first call to succeed will create the environment, everybody else gets the same thing. I think this must be made explicit. There is no possibility to create a separate instance with specific args and then have another one for general sharing.
2) If multiple calls are allowed (not recommended or needed) then there should be corresponding number of release calls. The notion of owned/not owned is simply non-existent.

I like the approach with the delegate.

Reply:
Thanks for this very useful information, I'd mirrored a lot of this approach from OrtAllocator but there is clearly good reason a similar pattern cannot apply here.
The challenge I'd been finding with this is that the logging function pointer cannot be set / modified after OrtEnv created.
I can wind this back and use a static OrtLoggingFunction delegate instead (picked up on singleton creation) as a simpler starting point, then proceed from there?

Reply:
This issue is known and came up before. I made an attempt to resolve it, but the PR got stalled due to other high priority issues.
[https://github.com/microsoft/onnxruntime/pull/14723](https://github.com/microsoft/onnxruntime/pull/14723)

Another common issue, is that folks forget to instantiate it first before doing anything else. Here is an attempt to fix it:
[https://github.com/microsoft/onnxruntime/pull/11593](https://github.com/microsoft/onnxruntime/pull/11593)
However, then everyone would need to be aware of this somewhat hidden default.

Reply:
I will spend some time on this.